docs: Created migration guide from SSD -> Distributed (#16389)
grafana/loki: docs/sources/setup/migrate/ssd-to-distributed/_index.md

---

> title: Migrate from SSD to Distributed

Reviewer: ```suggestion
title: Migrate from SSD to distributed
```

> menuTitle: Migrate from SSD to Distributed

Reviewer: ```suggestion
menuTitle: Migrate from SSD to distributed
```

description: Migration guide from migrating from simple scalable deployment to a distributed microservices deployment.
weight: 300
keywords:
  - migrate
  - distributed
  - ssd
---

> # Migrate from SSD to Distributed

Reviewer: ```suggestion
# Migrate from SSD to distributed
```

This guide provides instructions for migrating from a [simple scalable deployment (SSD)](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#simple-scalable) to a [distributed deployment](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#microservices-mode) of Loki. Before starting the migration, make sure you have read the [considerations](#considerations) section.

## Considerations

Migrating from a simple scalable deployment to a distributed deployment with zero downtime is possible but requires careful planning. The following considerations should be taken into account:

1. **Helm Deployment:** This guide assumes that you have deploying Loki using Helm. Other migration methods are possible but are not covered in this guide.

> 1. **K8s Resources:** This migration method requires you to spin up distributed Loki pods before shutting down the SSD pods. This means that you need to have enough resources in your Kubernetes cluster to run both the SSD and distributed Loki pods at the same time.

Reviewer: ```suggestion
1. **Kubernetes Resources:** This migration method requires you to spin up distributed Loki pods before shutting down the SSD pods. This means that you need to have enough resources in your Kubernetes cluster to run both the SSD and distributed Loki pods at the same time.
```

> 1. **Data:**: No changes are required to your underlying data storage. Although data loss or corruption is unlikely, it is always recommended to back up your data before starting the migration process. If you are using a cloud provider you can take a snapshot/backup.

Reviewer: ```suggestion
1. **Data:** No changes are required to your underlying data storage. Although data loss or corruption is unlikely, it is always recommended to back up your data before starting the migration process. If you are using a cloud provider you can take a snapshot/backup.
```

> 1. **Configuration:**: We do not account for all configuration paramters in this guide. We only cover the parameters that need to be changed. Other paramters can remain the same. **However:** if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarly needed for the Grafana drilldown logs feature.

Reviewer 2: ```suggestion
1. **Configuration:**: We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other paramters can remain the same. **However:** if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarly needed for the Grafana drilldown logs feature.
```

> 1. **Configuration:**: We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other paramters can remain the same. **However:** if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarly needed for the Grafana drilldown logs feature.

Reviewer: ```suggestion
1. **Configuration:** We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other parameters can remain the same. **However:** if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarily needed for the Grafana drilldown logs feature.
```

> 1. **Configuration:** We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other parameters can remain the same. **However:** if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarily needed for the Grafana drilldown logs feature.

Reviewer: ```suggestion
1. **Configuration:** We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other parameters can remain the same. **However**, if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarily needed for the Grafana drilldown logs feature.
```

> 2. **Zone Aware Ingesters:** This guide does not currently account for Zone Aware Ingesters. Our current recommendation is to either disable Zone Aware Ingesters or to consult the [Mimir migration guide](https://grafana.com/docs/helm-charts/mimir-distributed/latest/migration-guides/migrate-from-single-zone-with-helm/). Take note not all paramters are equivalent between Mimir and Loki. 

Reviewer: ```suggestion
1. **Zone Aware Ingesters:** This guide does not currently account for Zone Aware Ingesters. Our current recommendation is to either disable Zone Aware Ingesters or to consult the [Mimir migration guide](https://grafana.com/docs/helm-charts/mimir-distributed/latest/migration-guides/migrate-from-single-zone-with-helm/). Take note not all paramters are equivalent between Mimir and Loki. 
```

## Prerequisites

Before starting the migration process, make sure you have the following prerequisites:

1. Access to your Kubernetes cluster via `kubectl`.

> 1. Helm Installed

Reviewer: ```suggestion
1. Helm installed.
```

> ## SSD Example

Reviewer: ```suggestion
## SSD example
```

> ## SSD example

Reviewer: ```suggestion
## Example SSD deployment
```

         schema: v13
         index:
           prefix: loki_index_
           period: 24h
   storage_config:
     aws:
       region: eu-central-1
       bucketnames: aws-chunks-bucket
       s3forcepathstyle: false
   ingester:
       chunk_encoding: snappy
   ruler:
    enable_api: true
    storage:
      type: s3
      s3:
        region: eu-central-1
        bucketnames: aws-ruler-bucket
        s3forcepathstyle: false
      alertmanager_url: http://prom:9093
   querier:
      max_concurrent: 4

   storage:
      type: s3
      bucketNames:
        chunks: "aws-chunks-bucket"
        ruler: "aws-ruler-bucket"
      s3:
        region: eu-central-1

deploymentMode: SimpleScalable

# SSD
backend:
 replicas: 2
read:
 replicas: 3
write:
 replicas: 3

# Distributed Loki
ingester:
 replicas: 0
 zoneAwareReplication:
  enabled: false

querier:
 replicas: 0
 maxUnavailable: 0
queryFrontend:
 replicas: 0
 maxUnavailable: 0
queryScheduler:
 replicas: 0
distributor:
 replicas: 0
 maxUnavailable: 0
compactor:
 replicas: 0
indexGateway:
 replicas: 0
 maxUnavailable: 0
ruler:
 replicas: 0
 maxUnavailable: 0

# Single binary Loki
singleBinary:
 replicas: 0

minio:
 enabled: false
```

> {{< admonition type="note" >}}

Reviewer: I would move the note above the Yaml content, because some people will jump ahead and miss it.

> Note: that this example is only a reference on the parameters that need to be changed. There will be other parameters within your own config such as `limits_config`, `gateway`, `compactor`, etc. These can remain the same.

Reviewer: ```suggestion
This example is only a reference on the parameters that need to be changed. There will be other parameters within your own config such as `limits_config`, `gateway`, `compactor`, etc. These can remain the same.
```

{{< /admonition >}}

> ## Stage 1: Deploying the Loki Distributed Components

Reviewer: ```suggestion
## Stage 1: Deploying the Loki distributed components
```

> In this stage, we will deploy the distributed Loki components alongside the SSD components. We will also change the `deploymentMode` to `SimpleScalable<->Distributed`. The `SimpleScalable<->Distributed` migration mode allows for a zero-downtime transition between SimpleScalable and fully Distributed architectures. During migration, both deployment types run simultaneously, sharing the same object storage backend.

Reviewer: ```suggestion
In this stage, we will deploy the distributed Loki components alongside the SSD components. We will also change the `deploymentMode` to `SimpleScalable<->Distributed`. The `SimpleScalable<->Distributed` migration mode allows for a zero-downtime transition between Simple Scalable and fully Distributed architectures. During migration, both deployment types run simultaneously, sharing the same object storage backend.
```

> ```

Reviewer: This is formatted as a code sample with the copy/paste button.  But I think it's meant to be a table showing what components you're running in each deployment mode?  Can we remove the code formatting if it's unnecessary and add some kind of lead in sentence that explains why this is included?

> SimpleScalable Components   ->    Distributed Components

Reviewer: ```suggestion
Simple Scalable Components   ->    Distributed Components
```

------------------------    ->    ---------------------

> write (StatefulSet)         ->    distributor + ingester
> read (Deployment)           ->    query-frontend + querier
> backend (StatefulSet)       ->    compactor + ruler + index-gateway

Reviewer: Let's match how we spell/capitalize these in the Components and Deployment modes topics.
```suggestion
write (StatefulSet)         ->    Distributor + Ingester
read (Deployment)           ->    Query Frontend + Querier
backend (StatefulSet)       ->    Compactor + Ruler + Index Gateway
```

```

> Request Routing:

Reviewer: ```suggestion
Request routing:
```

> Request routing:

Reviewer: In the generated output, this looks odd, like maybe it applies to the content above, instead of below.  It's also unclear why this appears at this point in the topic.  
```suggestion
How Loki handles request routing during the migration:
```

> The gateway (nginx) handles request routing based on endpoint type:

Reviewer: ```suggestion
The Gateway (nginx) handles request routing based on endpoint type:
```

1. Write Path (`loki/api/v1/push`):

>    * Initially routes to SimpleScalable write component

Reviewer: ```suggestion
   * Initially routes to Simple Scalable write component
```

>    * Gradually shifted to distributor component

Reviewer: ```suggestion
   * Gradually shifted to the Distributor
```

   * Both write paths share the same object storage, ensuring data consistency
2. Read Path (`/loki/api/v1/query`):

>    * Routes to either SimpleScalable read or Distributed query-frontend

Reviewer: ```suggestion
   * Routes to either Simple Scalable read or Distributed Query Frontend
```

   * Query results are consistent since both architectures read from same storage
3. Admin/Background Operations:
   * Compaction, retention, and rule evaluation handled by either backend or respective distributed components
   * Operations are coordinated through object storage locks

To start the migration process:  

> 1. Create a copy of your existing `values.yaml` file and name it `values-migration.yaml` (You can also make these changes within your current values file).

Reviewer: If they keep the same file name, then the command at line 254 would fail, wouldn't it?
```suggestion
1. Create a copy of your existing `values.yaml` file and name it `values-migration.yaml`.
```

    ```bash
    cp values.yaml values-migration.yaml
    ```

1. Next modify the following parameters; `deploymentMode`, `ingester` and components based on the annotations below.

   ```yaml
   ---
   loki:
     schemaConfig:
       configs:
         - from: "2024-04-01"
           store: tsdb
           object_store: s3
           schema: v13
           index:
             prefix: loki_index_
             period: 24h
     storage_config:

>        aws:

Reviewer: I don't think you've mentioned anywhere in the text that this example is specific for Amazon S3?  You might want to mention that if you're using Azure or GCS what you'd need to change in those files?  OR should it be very similar?

         region: eu-central-1
         bucketnames: aws-chunks-bucket
         s3forcepathstyle: false
     ingester:
       chunk_encoding: snappy
       # Add this to ingester; this will force ingesters to flush before shutting down
       wal:

>          flush_on_shutdown: true

Reviewer 2: Should this be removed after a finished migration?

           region: eu-central-1
           bucketnames: aws-ruler-bucket
           s3forcepathstyle: false
         alertmanager_url: http://prom:9093
     querier:
       max_concurrent: 4

     storage:
       type: s3
       bucketNames:
         chunks: "aws-chunks-bucket"
         ruler: "aws-ruler-bucket"
       s3:
         region: eu-central-1

   # Important: Make sure to change this to SimpleScalable<->Distributed
   deploymentMode: SimpleScalable<->Distributed

   # SSD
   backend:
     replicas: 2
   read:
     replicas: 3
   write:
     replicas: 3

   # Distributed Loki
   # Spin up the distributed components
   ingester:
     replicas: 3
     zoneAwareReplication:
       enabled: false
   querier:
     replicas: 3
     maxUnavailable: 0
   queryFrontend:
     replicas: 2
     maxUnavailable: 0
   queryScheduler:
     replicas: 2
   distributor:
     replicas: 2
     maxUnavailable: 0
   compactor:
     replicas: 1
   indexGateway:
     replicas: 2
     maxUnavailable: 0
   ruler:
     replicas: 1
     maxUnavailable: 0

   # Single binary Loki
   singleBinary:
     replicas: 0

   minio:
     enabled: false
   ```

   Here is a breakdown of the changes:

>    * `ingester.wal.flush_on_shutdown: true`: This will force the ingesters to flush before shutting down. This is important to ommitt data loss.

Reviewer: ```suggestion
   * `ingester.wal.flush_on_shutdown: true`: This will force the ingesters to flush before shutting down. This is important to prevent data loss.
```

   * `deploymentMode: SimpleScalable<->Distributed`: This will allow for the SSD and distributed components to run simultaneously.
   * Spin up all distributed components with the desired replicas.

1. Deploy the distributed components using the following command:

    ```bash
    helm upgrade --values values-migration.yaml loki grafana/loki -n loki 
    ```
   {{< admonition type="caution" >}}
   It is important to allow all components to fully spin up before proceeding to the next stage. You can check the status of the components using the following command:

    ```bash
    kubectl get pods -n loki
    ```
   Let all components reach the `Running` state before proceeding to the next stage.
   {{< /admonition >}}

> ## Stage 2: Transitioning to Distributed Components

Reviewer: ```suggestion
## Stage 2: Transitioning to Distributed components
```

> ## Stage 2: Transitioning to Distributed components

Reviewer: ```suggestion
## Stage 2: Transitioning to distributed components
```

The final stage of the migration involves transitioning all traffic to the distributed components. This is done by scaling down the SSD components and swapping the `deploymentMode` to `Distributed`. To do this:

> 1. Create a copy of `values-migration.yaml` and name it `values-distributed.yaml` (Once again this can all be done in the original file).

Reviewer: Just noticing that there is not a command to apply this new file after it has been modified.

         alertmanager_url: http://prom:9093
     querier:
       max_concurrent: 4

     storage:
       type: s3
       bucketNames:
         chunks: "aws-chunks-bucket"
         ruler: "aws-ruler-bucket"
       s3:
         region: eu-central-1

   # Important: Make sure to change this to Distributed
   deploymentMode: Distributed

   # SSD
   # Scale down the SSD components
   backend:
     replicas: 0
   read:
     replicas: 0
   write:
     replicas: 0

   # Distributed Loki
   ingester:
     replicas: 3
     zoneAwareReplication:
       enabled: false
   querier:
     replicas: 3
     maxUnavailable: 0
   queryFrontend:
     replicas: 2
     maxUnavailable: 0
   queryScheduler:
     replicas: 2
   distributor:
     replicas: 2
     maxUnavailable: 0
   compactor:
     replicas: 1
   indexGateway:
     replicas: 2
     maxUnavailable: 0
   ruler:
     replicas: 1
     maxUnavailable: 0

   # Single binary Loki
   singleBinary:
     replicas: 0

   minio:
     enabled: false
   ```
  Here is a breakdown of the changes:
  * `deploymentMode: Distributed`: This will allow for the distributed components to run in isolation.
  * Scale down all SSD components to `0`.

> ## Migrating from SSD to Single Binary

Reviewer: I wouldn't expect to find this in a topic where the heading is "Migrate from SSD to distributed."  I think we can delete this section.

> If you are looking to migrate from SSD to a single binary deployment the process is similar to the migration from SSD to Distributed. The primary differnce is that you will use `SingleBinary<->SimpleScalable` as your `deploymentMode`. Before moving to `SingleBinary`.

Reviewer: ```suggestion
If you are looking to migrate from SSD to a single binary deployment the process is similar to the migration from SSD to Distributed. The primary difference is that you will use `SingleBinary<->SimpleScalable` as your `deploymentMode`. Before moving to `SingleBinary`.
```

> ## Summary

Reviewer: Not sure that we need the summary either?  Maybe move it to the top and revise to use as an overview?

{{< admonition type="note" >}}
In this guide, an AWS deployment is used as an example. However, the migration process is mirrored for other cloud providers. This is due to the fact that no changes are required to the underlying data storage.
{{< /admonition >}}

## Considerations

Migrating from a simple scalable deployment to a distributed deployment with zero downtime is possible but requires careful planning. The following considerations should be taken into account:

1. **Helm Deployment:** This guide assumes that you have deploying Loki using Helm. Other migration methods are possible but are not covered in this guide.
1. **Kubernetes Resources:** This migration method requires you to spin up distributed Loki pods before shutting down the SSD pods. This means that you need to have enough resources in your Kubernetes cluster to run both the SSD and distributed Loki pods at the same time.
1. **Data:** No changes are required to your underlying data storage. Although data loss or corruption is unlikely, it is always recommended to back up your data before starting the migration process. If you are using a cloud provider you can take a snapshot/backup.

> 1. **Configuration:** We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other parameters can remain the same. **However**, if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarily needed for the Grafana drilldown logs feature.

Reviewer: ```suggestion
1. **Configuration:** We do not account for all configuration parameters in this guide. We only cover the parameters that need to be changed. Other parameters can remain the same. **However**, if `pattern_ingesters=true` you will need to spin up `patternIngesters` before shutting down the SSD ingesters. This is primarily needed for the Grafana Logs Drilldown feature.
```

> 1. **Zone Aware Ingesters:** This guide does not currently account for Zone Aware Ingesters. Our current recommendation is to either disable Zone Aware Ingesters or to consult the [Mimir migration guide](https://grafana.com/docs/helm-charts/mimir-distributed/latest/migration-guides/migrate-from-single-zone-with-helm/). Take note not all paramters are equivalent between Mimir and Loki. 

Reviewer: ```suggestion
1. **Zone Aware Ingesters:** This guide does not currently account for Zone Aware Ingesters. Our current recommendation is to either disable Zone Aware Ingesters or to consult the [Mimir migration guide](https://grafana.com/docs/helm-charts/mimir-distributed/latest/migration-guides/migrate-from-single-zone-with-helm/). Take note, not all parameters are equivalent between Mimir and Loki. 
```

           period: 24h
   storage_config:
     aws:
       region: eu-central-1
       bucketnames: aws-chunks-bucket
       s3forcepathstyle: false
   ingester:
       chunk_encoding: snappy
   ruler:
    enable_api: true
    storage:
      type: s3
      s3:
        region: eu-central-1
        bucketnames: aws-ruler-bucket
        s3forcepathstyle: false
      alertmanager_url: http://prom:9093
   querier:
      max_concurrent: 4

   storage:
      type: s3
      bucketNames:
        chunks: "aws-chunks-bucket"
        ruler: "aws-ruler-bucket"
      s3:
        region: eu-central-1

deploymentMode: SimpleScalable

# SSD
backend:
 replicas: 2
read:
 replicas: 3
write:
 replicas: 3

# Distributed Loki
ingester:
 replicas: 0
 zoneAwareReplication:
  enabled: false

querier:
 replicas: 0
 maxUnavailable: 0
queryFrontend:
 replicas: 0
 maxUnavailable: 0
queryScheduler:
 replicas: 0
distributor:
 replicas: 0
 maxUnavailable: 0
compactor:
 replicas: 0
indexGateway:
 replicas: 0
 maxUnavailable: 0
ruler:
 replicas: 0
 maxUnavailable: 0

# Single binary Loki
singleBinary:
 replicas: 0

minio:
 enabled: false
```

## Stage 1: Deploying the Loki distributed components

> In this stage, we will deploy the distributed Loki components alongside the SSD components. We will also change the `deploymentMode` to `SimpleScalable<->Distributed`. The `SimpleScalable<->Distributed` migration mode allows for a zero-downtime transition between Simple Scalable and fully Distributed architectures. During migration, both deployment types run simultaneously, sharing the same object storage backend.

Reviewer: ```suggestion
In this stage, we will deploy the distributed Loki components alongside the SSD components. We will also change the `deploymentMode` to `SimpleScalable<->Distributed`. The `SimpleScalable<->Distributed` migration mode allows for a zero-downtime transition between Simple Scalable and fully distributed architectures. During migration, both deployment types run simultaneously, sharing the same object storage backend.
```

> The the following table outlines which components take over the responsibilities of the SSD components:

Reviewer: ```suggestion
The following table outlines which components take over the responsibilities of the SSD components:
```

Simple Scalable Components |  Distributed Components
---------------------------|--------------------------------
write (Deployment)         |   Distributor + Ingester
read (StatefulSet)         |   Query Frontend + Querier
backend (StatefulSet)      |   Compactor + Ruler + Index Gateway

How Loki handles request routing during the migration:

The Gateway (nginx) handles request routing based on endpoint type:
1. Write Path (`loki/api/v1/push`):
   * Initially routes to Simple Scalable write component
   * Gradually shifted to the Distributor
   * Both write paths share the same object storage, ensuring data consistency
2. Read Path (`/loki/api/v1/query`):

>    * Routes to either Simple Scalable read or Distributed Query Frontend

Reviewer: ```suggestion
   * Routes to either Simple Scalable read or distributed Query Frontend
```

   ```

   Here is a breakdown of the changes:
   * `ingester.wal.flush_on_shutdown: true`: This will force the ingesters to flush before shutting down. This is important to prevent data loss.
   * `deploymentMode: SimpleScalable<->Distributed`: This will allow for the SSD and distributed components to run simultaneously.
   * Spin up all distributed components with the desired replicas.

1. Deploy the distributed components using the following command:

    ```bash
    helm upgrade --values values-migration.yaml loki grafana/loki -n loki 
    ```
   {{< admonition type="caution" >}}
   It is important to allow all components to fully spin up before proceeding to the next stage. You can check the status of the components using the following command:

    ```bash
    kubectl get pods -n loki
    ```
   Let all components reach the `Running` state before proceeding to the next stage.
   {{< /admonition >}}

## Stage 2: Transitioning to distributed components

The final stage of the migration involves transitioning all traffic to the distributed components. This is done by scaling down the SSD components and swapping the `deploymentMode` to `Distributed`. To do this:

> 1. Create a copy of `values-migration.yaml` and name it `values-distributed.yaml` (Once again this can all be done in the original file).

Reviewer: ```suggestion
1. Create a copy of `values-migration.yaml` and name it `values-distributed.yaml`.
```

     ruler:
       enable_api: true
       storage:
         type: s3
         s3:
           region: eu-central-1
           bucketnames: aws-ruler-bucket
           s3forcepathstyle: false
         alertmanager_url: http://prom:9093
     querier:
       max_concurrent: 4

     storage:
       type: s3
       bucketNames:
         chunks: "aws-chunks-bucket"
         ruler: "aws-ruler-bucket"
       s3:
         region: eu-central-1

   # Important: Make sure to change this to Distributed
   deploymentMode: Distributed

   # SSD
   # Scale down the SSD components
   backend:
     replicas: 0
   read:
     replicas: 0
   write:
     replicas: 0

   # Distributed Loki
   ingester:
     replicas: 3
     zoneAwareReplication:
       enabled: false
   querier:
     replicas: 3
     maxUnavailable: 0
   queryFrontend:
     replicas: 2
     maxUnavailable: 0
   queryScheduler:
     replicas: 2
   distributor:
     replicas: 2
     maxUnavailable: 0
   compactor:
     replicas: 1
   indexGateway:
     replicas: 2
     maxUnavailable: 0
   ruler:
     replicas: 1
     maxUnavailable: 0

   # Single binary Loki
   singleBinary:
     replicas: 0

   minio:
     enabled: false
   ```

>   Here is a breakdown of the changes:
>   * `deploymentMode: Distributed`: This will allow for the distributed components to run in isolation.
>   * Scale down all SSD components to `0`.

Reviewer: ```suggestion
     Here is a breakdown of the changes:
     * `deploymentMode: Distributed`: This will allow for the distributed components to run in isolation.
     * Scale down all SSD components to `0`.
```

1. Deploy the final configuration using the following command:

    ```bash
    helm upgrade --values values-distributed.yaml loki grafana/loki -n loki 
    ```

1. Once the deployment is complete, you can verify that all components are running using the following command:

    ```bash
    kubectl get pods -n loki
    ```

You should see all distributed components running and the SSD compontents have now been removed.

## What's next?

> Loki in distributed mode is inherently more complex than SSD mode. It is recommended to meta-minitor your Loki deployment to ensure that everything is running smoothly. You can do this by following the [meta-monitoring guide](https://grafana.com/docs/loki/latest/operations/meta-monitoring/). 

Reviewer: ```suggestion
Loki in distributed mode is inherently more complex than SSD mode. It is recommended to meta-monitor your Loki deployment to ensure that everything is running smoothly. You can do this by following the [meta-monitoring guide](https://grafana.com/docs/loki/<LOKI_VERSION>/operations/meta-monitoring/). 
```
